File: src-tauri/src/commands/ingest.rs
use std::io::Cursor;
use std::path::{Path, PathBuf};

use image::io::Reader as ImageReader;
use tauri::{AppHandle, Emitter, State};
use tokio::task::spawn_blocking;

use crate::error::{AppError, Result};
use crate::ingest::{process_pdf, IngestionProgress, IngestionResult};
use crate::state::AppState;

/// Ingest a PDF file into the database
#[tauri::command]
pub async fn ingest_pdf(
    file_path: String,
    title: Option<String>,
    author: Option<String>,
    app_handle: AppHandle,
    state: State<'_, AppState>,
) -> Result<IngestionResult> {
    let pool = state.get_pool().await.ok_or(AppError::NotConnected)?;

    let path = PathBuf::from(&file_path);

    let filename = path
        .file_name()
        .and_then(|n| n.to_str())
        .map(|s| s.to_string());

    let _ = app_handle.emit("ingestion-progress", IngestionProgress::reading(0));

    let pdf_result = spawn_blocking(move || process_pdf(&path))
        .await
        .map_err(|e| AppError::PdfError(format!("Task join error: {}", e)))??;

    let page_count = pdf_result.page_count;

    let final_title = title.or(pdf_result.metadata.title);
    let final_author = author.or(pdf_result.metadata.author);

    let _ = app_handle.emit("ingestion-progress", IngestionProgress::reading(page_count));

    let mut tx = pool.begin().await?;

    let file_id: i64 =
        sqlx::query_scalar(r#"INSERT INTO file (type, path) VALUES ('raw', $1) RETURNING id"#)
            .bind(&file_path)
            .fetch_one(&mut *tx)
            .await?;

    let document_id: i64 = sqlx::query_scalar(
        r#"INSERT INTO document (path, filename, author, title) VALUES ($1, $2, $3, $4) RETURNING id"#,
    )
    .bind(file_id)
    .bind(&filename)
    .bind(&final_author)
    .bind(&final_title)
    .fetch_one(&mut *tx)
    .await?;

    let mimetype = "image/png".to_string();

    for (page_idx, png_bytes) in pdf_result.pages.into_iter().enumerate() {
        let page_metadata = serde_json::json!({"source_path": file_path});
        let page_id: i64 = sqlx::query_scalar(
            r#"INSERT INTO page (page_num, document_id, image_contents, mimetype, page_metadata)
               VALUES ($1, $2, NULL, $3, $4) RETURNING id"#,
        )
        .bind((page_idx + 1) as i32)
        .bind(document_id)
        .bind(&mimetype)
        .bind(&page_metadata)
        .fetch_one(&mut *tx)
        .await?;

        sqlx::query_scalar::<_, i64>(
            r#"INSERT INTO image_chunk (parent_page, contents, mimetype)
               VALUES ($1, $2, $3) RETURNING id"#,
        )
        .bind(page_id)
        .bind(&png_bytes)
        .bind(&mimetype)
        .fetch_one(&mut *tx)
        .await?;
    }

    tx.commit().await?;

    let _ = app_handle.emit(
        "ingestion-progress",
        IngestionProgress::complete(page_count),
    );

    Ok(IngestionResult {
        file_id,
        document_id,
        page_count,
        image_chunk_count: page_count,
    })
}

/// Load an image file and convert it to PNG bytes
fn load_image_as_png(path: &Path) -> Result<Vec<u8>> {
    let img = ImageReader::open(path)
        .map_err(|e| AppError::ImageError(format!("Failed to open image: {}", e)))?
        .decode()
        .map_err(|e| AppError::ImageError(format!("Failed to decode image: {}", e)))?;

    let mut png_bytes = Vec::new();
    img.write_to(&mut Cursor::new(&mut png_bytes), image::ImageFormat::Png)
        .map_err(|e| AppError::ImageError(format!("Failed to encode as PNG: {}", e)))?;

    Ok(png_bytes)
}

/// Ingest multiple image files into the database as a single document
#[tauri::command]
pub async fn ingest_images(
    file_paths: Vec<String>,
    title: String,
    app_handle: AppHandle,
    state: State<'_, AppState>,
) -> Result<IngestionResult> {
    if file_paths.is_empty() {
        return Err(AppError::ImageError("No files provided".to_string()));
    }

    if title.trim().is_empty() {
        return Err(AppError::ImageError("Title is required".to_string()));
    }

    let pool = state.get_pool().await.ok_or(AppError::NotConnected)?;

    let total_images = file_paths.len() as i32;

    let _ = app_handle.emit(
        "ingestion-progress",
        IngestionProgress::reading(total_images),
    );

    let paths: Vec<PathBuf> = file_paths.iter().map(PathBuf::from).collect();
    for path in &paths {
        if !path.exists() {
            return Err(AppError::ImageError(format!(
                "File not found: {}",
                path.display()
            )));
        }
    }

    let app_handle_clone = app_handle.clone();
    let image_data: Vec<Vec<u8>> = spawn_blocking(move || {
        let mut results = Vec::with_capacity(paths.len());
        for (idx, path) in paths.iter().enumerate() {
            let _ = app_handle_clone.emit(
                "ingestion-progress",
                IngestionProgress::rendering((idx + 1) as i32, total_images),
            );

            let png_bytes = load_image_as_png(path)?;
            results.push(png_bytes);
        }
        Ok::<_, AppError>(results)
    })
    .await
    .map_err(|e| AppError::ImageError(format!("Task join error: {}", e)))??;

    let mut tx = pool.begin().await?;

    let document_id: i64 = sqlx::query_scalar(
        r#"INSERT INTO document (path, filename, title) VALUES (NULL, NULL, $1) RETURNING id"#,
    )
    .bind(&title)
    .fetch_one(&mut *tx)
    .await?;

    let mimetype = "image/png".to_string();

    for (page_idx, png_bytes) in image_data.into_iter().enumerate() {
        let page_metadata = serde_json::json!({"source_path": file_paths[page_idx]});
        let page_id: i64 = sqlx::query_scalar(
            r#"INSERT INTO page (page_num, document_id, image_contents, mimetype, page_metadata)
               VALUES ($1, $2, NULL, $3, $4) RETURNING id"#,
        )
        .bind((page_idx + 1) as i32)
        .bind(document_id)
        .bind(&mimetype)
        .bind(&page_metadata)
        .fetch_one(&mut *tx)
        .await?;

        sqlx::query_scalar::<_, i64>(
            r#"INSERT INTO image_chunk (parent_page, contents, mimetype)
               VALUES ($1, $2, $3) RETURNING id"#,
        )
        .bind(page_id)
        .bind(&png_bytes)
        .bind(&mimetype)
        .fetch_one(&mut *tx)
        .await?;
    }

    tx.commit().await?;

    let _ = app_handle.emit(
        "ingestion-progress",
        IngestionProgress::complete(total_images),
    );

    Ok(IngestionResult {
        file_id: 0,
        document_id,
        page_count: total_images,
        image_chunk_count: total_images,
    })
}

/// Get supported file formats for ingestion
#[tauri::command]
pub fn get_supported_formats() -> Vec<&'static str> {
    vec!["pdf", "png", "jpg", "jpeg", "webp"]
}

File: src/stores/ingest.ts
import { acceptHMRUpdate, defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { invoke } from '@tauri-apps/api/core'
import { listen, type UnlistenFn } from '@tauri-apps/api/event'

export interface IngestionProgress {
  current_page: number
  total_pages: number
  phase: 'Reading' | 'Rendering' | 'Complete' | 'Failed'
  message: string
}

export interface IngestionResult {
  file_id: number
  document_id: number
  page_count: number
  image_chunk_count: number
}

export const useIngestStore = defineStore('ingest', () => {
  const isIngesting = ref(false)
  const progress = ref<IngestionProgress | null>(null)
  const error = ref<string | null>(null)
  const lastResult = ref<IngestionResult | null>(null)

  let unlistenProgress: UnlistenFn | null = null

  const progressPercent = computed(() => {
    if (!progress.value || progress.value.total_pages === 0) return 0
    return Math.round((progress.value.current_page / progress.value.total_pages) * 100)
  })

  const isComplete = computed(() => progress.value?.phase === 'Complete')
  const isFailed = computed(() => progress.value?.phase === 'Failed')

  async function startListening() {
    if (unlistenProgress) {
      unlistenProgress()
    }

    unlistenProgress = await listen<IngestionProgress>('ingestion-progress', (event) => {
      progress.value = event.payload
      if (event.payload.phase === 'Failed') {
        error.value = event.payload.message
      }
    })
  }

  function stopListening() {
    if (unlistenProgress) {
      unlistenProgress()
      unlistenProgress = null
    }
  }

  async function ingestPdf(filePath: string, title?: string, author?: string): Promise<IngestionResult | null> {
    isIngesting.value = true
    progress.value = null
    error.value = null
    lastResult.value = null

    try {
      await startListening()

      const result = await invoke<IngestionResult>('ingest_pdf', {
        filePath,
        title: title || null,
        author: author || null,
      })

      lastResult.value = result
      return result
    } catch (err) {
      error.value = err instanceof Error ? err.message : String(err)
      progress.value = {
        current_page: 0,
        total_pages: 0,
        phase: 'Failed',
        message: error.value,
      }
      return null
    } finally {
      isIngesting.value = false
      stopListening()
    }
  }

  async function ingestImages(filePaths: string[], title: string): Promise<IngestionResult | null> {
    isIngesting.value = true
    progress.value = null
    error.value = null
    lastResult.value = null

    try {
      await startListening()

      const result = await invoke<IngestionResult>('ingest_images', {
        filePaths,
        title,
      })

      lastResult.value = result
      return result
    } catch (err) {
      error.value = err instanceof Error ? err.message : String(err)
      progress.value = {
        current_page: 0,
        total_pages: 0,
        phase: 'Failed',
        message: error.value,
      }
      return null
    } finally {
      isIngesting.value = false
      stopListening()
    }
  }

  async function getSupportedFormats(): Promise<string[]> {
    try {
      return await invoke<string[]>('get_supported_formats')
    } catch {
      return ['pdf']
    }
  }

  function reset() {
    isIngesting.value = false
    progress.value = null
    error.value = null
    lastResult.value = null
  }

  return {
    isIngesting,
    progress,
    error,
    lastResult,
    progressPercent,
    isComplete,
    isFailed,
    ingestPdf,
    ingestImages,
    getSupportedFormats,
    reset,
  }
})

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useIngestStore, import.meta.hot))
}
